### PR TITLE
Move memory and prompt calculations out of the Chat interface

### PR DESCRIPTION
**Pull request by raulraja**
Simplifying the `Chat` interface implementation creates two new internal objects for handling concerns related to Memory and Prompt model token calculations.

### CODE REVIEW

**Review by jackcviers — APPROVED** [review]

File: core/src/commonMain/kotlin/com/xebia/functional/xef/llm/PromptCalculator.kt [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+        Prompt { +assistant(ctxTruncated) }.messages
+      } else {
+        emptyList()
+      }
+
+    return prompt.copy(messages = contextAllowed + historyAllowed + prompt.messages)
+  }
+
+  fun messagesFromMemory(memories: List<Memory>): List<Message> = memories.map { it.content }
```
Comment (javipacheco) [inline core/src/commonMain/kotlin/com/xebia/functional/xef/llm/PromptCalculator.kt:54]: This method should be `private`

File: core/src/commonMain/kotlin/com/xebia/functional/xef/llm/PromptCalculator.kt [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+  fun calculateMessagesFromHistory(chat: Chat, memories: List<Memory>, maxHistoryTokens: Int) =
```
Comment (javipacheco) [inline core/src/commonMain/kotlin/com/xebia/functional/xef/llm/PromptCalculator.kt:56]: This method should be `private`

File: core/src/commonMain/kotlin/com/xebia/functional/xef/llm/PromptCalculator.kt [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+              acc
+            } else {
+              Pair(acc.first + pair.second, acc.second + pair.first)
+            }
+          }
+        totalTokenWithMessages.second.reversed()
+      }
+    } else emptyList()
+
+  fun calculateMaxContextTokens(prompt: Prompt, remainingTokensForContexts: Int): Int {
```
Comment (javipacheco) [inline core/src/commonMain/kotlin/com/xebia/functional/xef/llm/PromptCalculator.kt:80]: This method should be `private`

File: core/src/commonMain/kotlin/com/xebia/functional/xef/llm/PromptCalculator.kt [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    val contextPercent = prompt.configuration.messagePolicy.contextPercent
+    val maxContextTokens = (remainingTokensForContexts * contextPercent) / 100
+    return maxContextTokens
+  }
+
+  fun calculateMaxHistoryTokens(prompt: Prompt, remainingTokensForContexts: Int): Int {
```
Comment (javipacheco) [inline core/src/commonMain/kotlin/com/xebia/functional/xef/llm/PromptCalculator.kt:86]: This method should be `private`

File: core/src/commonMain/kotlin/com/xebia/functional/xef/llm/PromptCalculator.kt [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    val historyPercent = prompt.configuration.messagePolicy.historyPercent
+    val maxHistoryTokens = (remainingTokensForContexts * historyPercent) / 100
+    return maxHistoryTokens
+  }
+
+  fun calculateRemainingTokensForContext(chat: Chat, prompt: Prompt): Int {
```
Comment (javipacheco) [inline core/src/commonMain/kotlin/com/xebia/functional/xef/llm/PromptCalculator.kt:92]: This method should be `private`

File: core/src/commonMain/kotlin/com/xebia/functional/xef/llm/PromptCalculator.kt [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+    if (messagesTokens >= remainingTokens) {
+      throw AIError.PromptExceedsMaxRemainingTokenLength(messagesTokens, remainingTokens)
+    }
+
+    val remainingTokensForContexts = remainingTokens - messagesTokens
+    return remainingTokensForContexts
+  }
+
+  private suspend fun Conversation.memories(limitTokens: Int): List<Memory> {
```
Comment (javipacheco) [inline core/src/commonMain/kotlin/com/xebia/functional/xef/llm/PromptCalculator.kt:106]: This method should be `private`

**Review by javipacheco — DISMISSED** [review]
Great job 🚀 

Just a minor comments if you want to change them

**Review by javipacheco — APPROVED** [review]
🚀

**Pull request merged**